File: src/InputData.cpp
/*
 * Copyright 2020 Maximilian Azendorf
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "InputData.h"

#include <algorithm>
#include <climits>

#include "Util.h"
#include "UnionFind.h"
#include "Constraints.h"
#include "input/InputException.h"

int InputData::preference_after(int preference) const
{
    // TODO: Maybe do this with binary search.
    for(int i = 0; i < _preferenceLevels.size(); i++)
    {
        if(_preferenceLevels[i] > preference)
        {
            return _preferenceLevels[i];
        }
    }

    return INT_MAX;
}

vector<Constraint> const& InputData::scheduling_constraints() const
{
    return _schedulingConstraints;
}

vector<Constraint> const& InputData::scheduling_constraints(int choiceId) const
{
    return _choiceConstraintMap.at(choiceId);
}

vector<Constraint> const& InputData::assignment_constraints() const
{
    return _assignmentConstraints;
}

vector<Constraint> const& InputData::assignment_constraints(int chooserId) const
{
    return _chooserConstraintMap.at(chooserId);
}

vector<ChoiceData> const& InputData::choices() const
{
    return _choices;
}

vector<ChooserData> const& InputData::choosers() const
{
    return _choosers;
}

vector<SlotData> const& InputData::slots() const
{
    return _slots;
}

ChoiceData const& InputData::choice(int index) const
{
    return _choices[index];
}

ChooserData const& InputData::chooser(int index) const
{
    return _choosers[index];
}

SlotData const& InputData::slot(int index) const
{
    return _slots[index];
}

vector<vector<int>> const& InputData::dependent_choice_groups() const
{
    return _dependentChoiceGroups;
}

vector<int> const& InputData::preference_levels() const
{
    return _preferenceLevels;
}

int InputData::max_preference() const
{
    return _maxPreference;
}

int InputData::choice_count() const
{
    return _choices.size();
}

int InputData::chooser_count() const
{
    return _choosers.size();
}

int InputData::slot_count() const
{
    return _slots.size();
}

File: src/input_data.rs
use std::collections::BTreeMap;

use crate::{ChoiceData, ChooserData, Constraint, SlotData};

/// Fully parsed and preprocessed input data.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InputData {
    pub(crate) choices: Vec<ChoiceData>,
    pub(crate) choosers: Vec<ChooserData>,
    pub(crate) slots: Vec<SlotData>,
    pub(crate) scheduling_constraints: Vec<Constraint>,
    pub(crate) assignment_constraints: Vec<Constraint>,
    pub(crate) dependent_choice_groups: Vec<Vec<usize>>,
    pub(crate) preference_levels: Vec<i32>,
    pub(crate) max_preference: i32,
    pub(crate) choice_constraint_map: BTreeMap<usize, Vec<Constraint>>,
    pub(crate) chooser_constraint_map: BTreeMap<usize, Vec<Constraint>>,
}

impl InputData {
    pub(crate) const GENERATED_PREFIX: &str = "~";
    pub(crate) const GENERATED_SLOT_NAME: &str = "Generated Slot";

    pub(crate) fn preference_after(&self, preference: i32) -> i32 {
        self.preference_levels
            .iter()
            .copied()
            .find(|&level| level > preference)
            .unwrap_or(i32::MAX)
    }

    pub(crate) fn scheduling_constraints_for(&self, choice: usize) -> &[Constraint] {
        self.choice_constraint_map
            .get(&choice)
            .map_or(self.scheduling_constraints.as_slice(), Vec::as_slice)
    }

    /// Returns all scheduling constraints in the input.
    #[must_use]
    pub fn scheduling_constraints(&self) -> &[Constraint] {
        &self.scheduling_constraints
    }

    /// Returns all assignment constraints in the input.
    #[must_use]
    pub fn assignment_constraints(&self) -> &[Constraint] {
        &self.assignment_constraints
    }

    /// Returns the maximum preference value occurring in the input.
    #[must_use]
    pub fn max_preference(&self) -> i32 {
        self.max_preference
    }

    /// Returns the number of choices in the input.
    #[must_use]
    pub fn choice_count(&self) -> usize {
        self.choices.len()
    }

    /// Returns the parsed choices.
    #[must_use]
    pub fn choices(&self) -> &[ChoiceData] {
        &self.choices
    }

    /// Returns the number of choosers in the input.
    #[must_use]
    pub fn chooser_count(&self) -> usize {
        self.choosers.len()
    }

    /// Returns the parsed choosers.
    #[must_use]
    pub fn choosers(&self) -> &[ChooserData] {
        &self.choosers
    }

    /// Returns the number of slots in the input.
    #[must_use]
    pub fn slot_count(&self) -> usize {
        self.slots.len()
    }

    /// Returns the parsed slots.
    #[must_use]
    pub fn slots(&self) -> &[SlotData] {
        &self.slots
    }
}
